New /dev/crypto engine

levitte opened the pull request:
Add a new devcrypto engine.  This has been implemented based on [cryptodev-linux](https://github.com/cryptodev-linux/cryptodev-linux) and then adjusted to work on FreeBSD 8.4.  It is assumed to build on other BSD platforms as well (**testers welcome!**).

To enable on non-BSD platforms, add `-DOPENSSL_USE_DEVCRYPTO` to your configuration.

Assymetric operations are *not* supported due to the deplaurable documentation.

levitte committed:
Add new /dev/crypto engine

Based on cryptodev-linux

levitte committed:
Adapt for BSD cryptodev.h differences

The BSD cryptodev.h doesn't have things like COP_FLAG_WRITE_IV and
COP_FLAG_UPDATE.  In that case, we need to implement that
functionality ourselves.

levitte committed:
Comment on the lack of documentation for asymmetric ciphers

levitte committed:
Document the added devcrypto engine in CHANGES

levitte commented:
Maybe this should be enabled through the usual `enable-{name}` mechanism?

richsalz commented:
Yes probably.  And also disable-cryptodev.

levitte commented:
Redone to use `enable-devcryptoeng` / `disable-devcryptoeng`.  Depends on #3745

tmshort commented:
Shouldn't it be 'no-devcryptoeng' ?

levitte commented:
@tmshort, `disable-foo` and `no-foo` are synonyms for any `foo`

levitte commented:
Ping?

mattcaswell commented on crypto/engine/eng_devcrypto.c:
```diff
@@ -0,0 +1,643 @@
+ */
+static int known_cipher_nids[nelem(cipher_data)];
+static int known_cipher_nids_amount = -1; /* -1 indicates not yet initialised */
+static EVP_CIPHER *known_cipher_methods[nelem(cipher_data)] = { 0, };
```
NULL not 0?

levitte replied:
Okie

mattcaswell commented on crypto/engine/eng_devcrypto.c:
```diff
@@ -0,0 +1,643 @@
+{
+    size_t i = get_cipher_data_index(nid);
+
+    if (known_cipher_methods[i] == NULL) {
```
Is there a multi-threaded issue here?

mattcaswell commented on crypto/engine/eng_devcrypto.c:
```diff
@@ -0,0 +1,643 @@
+static int devcrypto_ciphers(ENGINE *e, const EVP_CIPHER **cipher,
+                             const int **nids, int nid)
+{
+    if (!cipher)
```
cipher == NULL

levitte replied:
Fixed

mattcaswell commented on crypto/engine/eng_devcrypto.c:
```diff
@@ -0,0 +1,643 @@
+
+static int get_cipher_nids(const int **nids)
+{
+    if (known_cipher_nids_amount < 0) {
```
multi-thread? Perhaps this should be called on engine load?

levitte replied:
Hmmm, good point.

levitte replied:
I've modified it to prepare the tables beforehand...  pushed it just now

mattcaswell commented on crypto/engine/eng_devcrypto.c:
```diff
@@ -0,0 +1,643 @@
+#if !defined(CHECK_BSD_STYLE_MACROS) && defined(CRYPTO_SHA2_384)
+    { NID_sha384, 384 / 8, CRYPTO_SHA2_384 },
+#endif
+#if defined(CHECK_BSD_STYLE_MACROS) && defined(CRYPTO_SHA2_512)
```
Should this have a ! here like the others?

levitte replied:
Yes it should!  Good catch

mattcaswell commented on crypto/engine/eng_devcrypto.c:
```diff
@@ -0,0 +1,643 @@
+    if (!EVP_MD_CTX_test_flags(ctx, EVP_MD_CTX_FLAG_ONESHOT)) {
+        /*
+         * TODO: Should we limit the amount of bytes that can be processed
+         * with this method?  What if someone wants to hash a multi-GB file?
```
Is there no updatable hash state concept with devcrypto?

levitte replied:
Yes, there is...  if `COP_FLAG_UPDATE` and `COP_FLAG_FINAL` are defined.  Otherwise, no.

mattcaswell replied:
Right, ok. I hadn't quite absorbed that detail. I am minded to say that we don't allow multiple updates if the underlying engine doesn't support it.

levitte replied:
In that case, we're killing digests on all BSDs.  `COP_FLAG_UPDATE` and `COP_FLAG_FINAL` is cryptodev-linux' invention.  I find that a bit harsh

mattcaswell replied:
Well, only multiple update digests right? We can still support it if you call Init, Update, Final. Just not if you do Init, Update, Update, Final. Is this engine loaded by default on BSD? Or only on request? Did we support multiple updates on BSD previously?

levitte replied:
> Well, only multiple update digests right? We can still support it if you call Init, Update, Final.

Ah, so refuse unless `EVP_MD_CTX_FLAG_ONESHOT` is set.  Sure, that can be done, and should probably be done in `digest_init` rather than `digest_update`.

> Is this engine loaded by default on BSD? Or only on request?

It's enabled by default on BSD and loaded by default when enabled, yes.

> Did we support multiple updates on BSD previously?

I think the old `eng_cryptodev.c`.  Please confirm, I'm keeping my eyes away from that one.

mattcaswell replied:
> Ah, so refuse unless EVP_MD_CTX_FLAG_ONESHOT is set. Sure, that can be done, and should probably be done in digest_init rather than digest_update.

Well, no, that's not quite what I meant. EVP_MD_CTX_FLAG_ONESHOT, only gets set if you call EVP_Digest() - not if you call the sequence Init, Update, Final. Nevertheless we should still be able to support that sequence. Just treat all Updates as a oneshot, and fail if we detect a second update. That's assuming we confirm that the old cryptodev couldn't do it either. If it could then your way is probably the only way.

> I think the old eng_cryptodev.c did. Please confirm, I'm keeping my eyes away from that one.

Good idea. I think I should too as reviewer. Perhaps @richsalz can confirm?

mattcaswell reviewed (changes requested):
There don't seem to be any tests for this. Can we have some please? Also I made a few comments on the ciphers stuff which may also apply to the digests stuff. I didn't repeat myself.

levitte commented:
> There don't seem to be any tests for this. Can we have some please?

That can be done, but using it requires running the tests on BSD machines (I have no such access now) or on a Linux machine with `cryptodev.ko` loaded.  It will not be useful on Travis.

levitte commented on crypto/init.c:
```diff
@@ -560,6 +572,11 @@ int OPENSSL_init_crypto(uint64_t opts, const OPENSSL_INIT_SETTINGS *settings)
     if ((opts & OPENSSL_INIT_ENGINE_OPENSSL)
             && !RUN_ONCE(&engine_openssl, ossl_init_engine_openssl))
         return 0;
+# if !defined(OPENSSL_NO_HW) && !defined(OPENSSL_NO_DEVCRYPTOENG)
```
Good point

levitte replied:
Actually, no.  There's no hardware to disable, at least from our perspective, so `no-hw` shouldn't be considered at all here.
(as a matter of fact, `no-hw` should pretty much be a no-op for the moment)

levitte commented on crypto/engine/eng_devcrypto.c:
```diff
@@ -0,0 +1,643 @@
+
+#include "internal/engine.h"
+
+#define nelem(x) (sizeof(x)/sizeof(x[0]))
```
At some point, I wanted to make this as independent as possible.  Moot point now, I'll fix.

levitte commented on crypto/engine/eng_devcrypto.c:
```diff
@@ -0,0 +1,643 @@
+    int flags;
+    int devcryptoid;
+} cipher_data[] = {
+    { NID_des_cbc, 8, 8, 8, EVP_CIPH_CBC_MODE, CRYPTO_DES_CBC },
```
I should probably check for the presence of `CRYPTO_DES_CBC`, the same way I do with the other algos below...

levitte replied:
On second thought, no, I see no reason why.  If `/dev/crypto` supports DES operations, should we really obstruct that?

mattcaswell replied:
Well, I think many people would use no-des to explicitly disable DES with the thinking that "if I've used no-des no-one can possibly inadvertently use it by mistake". With this in place it would still be possible to use it.

levitte replied:
Hmmm, I'm ambivalent about this.  Also, I'm pretty sure the old `eng_cryptodev.c` didn't guard functionality this way, but am unsure if that should have an impact on this code.

levitte commented on crypto/engine/eng_devcrypto.c:
```diff
@@ -0,0 +1,643 @@
+     * are used.  If any other NID reaches this function, there's a grave
+     * coding error further down.
+     */
+    OPENSSL_die("Code that never should be reached", __FILE__, __LINE__);
```
Okie.

mattcaswell commented:
> That can be done, but using it requires running the tests on BSD machines (I have no such access now) or on a Linux machine with cryptodev.ko loaded. It will not be useful on Travis.

As a general rule we should be writing tests for all new functionality. In this instances the tests should be run via run checker.

levitte commented on crypto/engine/eng_devcrypto.c:
```diff
@@ -0,0 +1,643 @@
+    { NID_aes_256_ctr, 16, 256 / 8, 16, EVP_CIPH_CTR_MODE, CRYPTO_AES_CTR },
+#endif
+#if 0                            /* Not yet supported */
+    { NID_aes_128_xts, 16, 128 / 8, 16, EVP_CIPH_XTS_MODE, CRYPTO_AES_XTS },
```
Fixed

richsalz commented:
I agree with Matt about the no-des flag.

levitte commented:
> I agree with Matt about the no-des flag.

Ok, I hear you.  But in that case, we shouldn't stop at DES, the same guards should be there for all algos

richsalz commented:
Yes the guards should be there.  I’m not following this PR closely, just saw the DES comment fly by.

levitte commented:
Okie, done

levitte commented:
> > That can be done, but using it requires running the tests on BSD machines (I have no such access now) or on a Linux machine with cryptodev.ko loaded. It will not be useful on Travis.
> 
> As a general rule we should be writing tests for all new functionality. In this instances the tests should be run via run checker.

Thinking about it, I'm unsure what specific test to add, 'cause *all* the tests will test this engine if it's enabled.  So, errr, is it enough if I simply add an `enable-devcryptoeng` case in our run-checker?

mattcaswell commented:
> Thinking about it, I'm unsure what specific test to add, 'cause all the tests will test this engine if it's enabled. So, errr, is it enough if I simply add an enable-devcryptoeng case in our run-checker?

Good point. That might be sufficient. Unless we feel there are any edge cases worthy of testing that are specific to this engine?

levitte commented:
I can't think of any edge cases that won't be tested by something...

levitte commented:
As far as I can see, there's only one issue left, the multiple digest update one on BSDs.

mattcaswell commented:
@richsalz can you help with this? We'd like to check the requirements with regards to multiple digest "Update" calls on BSD. Did the previous implementation support this (which presumably required some internal caching of the input data, because cryptodev itself does not support it)? Richard understandably doesn't want to look at the original code, and I think it would be best for me to avoid it too as reviewer for this change.

richsalz commented:
I looked throught commit f39a5501ce69cab0c7282f5dcbf2b80d8ee259f2.  It appears that (a) digests are turned off because the code is wrapped with ```#ifdef USE_CRYPTODEV_DIGESTS``` which is never turned on. There's a big block comment with XXXX that says (paraphrasing) turn off for now, since we don't know which devices can efficiently will do which digests.

But the code does support the open/update/update/finish API.  Or it would, if it were turned on.

levitte commented:
I that with the old cryptodev engine was built around the idea of configuring with macros, i.e. something like this:

```
./config -DUSE_CRYPTODEV_DIGESTS
```

levitte commented:
... but ok, that comment about inefficient devices is another thing...  question is if it's still current.  After all, the old engine is quite old and things must have moved on since...

mattcaswell commented:
>I that with the old cryptodev engine was built around the idea of configuring with macros

Which implies that it is not on by default - which is probably the policy we should take.

levitte commented:
> Which implies that it is not on by default - which is probably the policy we should take.

I disagree...  we don't know what today's hardware is like.

richsalz commented:
It is hard to imagine pushing data across a syscall and/or a device boundary is more efficient than pure user-space, esp on modern hardware.

mattcaswell commented:
> I disagree... we don't know what today's hardware is like.

It's not the hardware that worries me. I don't like the idea of caching all the data to be digested until we've finished "updating". The whole point of the "Update" mechanism is to enable you to handle data streams (which could be quite large) and avoid the need to send all the data in one go. The solution that you currently have in place deliberately breaks that model with a potentially very large memory cost. Yes, we could implement some arbitrary upper limit - but that could just cause unexpected and difficult to debug breakage (i.e. your application works perfectly in dev and test - but breaks unexpectedly in prod). Not to mention the costs associated of moving all of that memory around, and the possible security risks if it leaks.

My worry was that if the old cryptodev supported it then probably we should too. But it seems it didn't, so I don't think we should worry about it. Just turn off digests by default and only allow them if you explicitly ask for them. Even in that case I'm thinking we should not allow the multi-update scenario (unless of course you're using a cryptodev that supports it).

levitte commented:
> Even in that case I'm thinking we should not allow the multi-update scenario (unless of course you're using a cryptodev that supports it).

You know, thinking of it all, I agree with this, and I'm more an more inclined to not care too much what the old engine did or didn't support (and especially if it didn't support digests at all by default, we really have no need to care!)

Re speed, there are all sorts of claims that neither digests or ciphers are helped very much by specialised hardware, these days.  However, there's also the claim that it's better to use kernel / hardware crypto for security, or possibly to use available cycles more efficiently.  Between these claims, I'm ready to throw my hands in the air and point out that the user / packager is at his/her liberty to enable or disable this feature at their leasure.

levitte committed:
Add the common error ERR_R_OPERATION_FAIL

levitte committed:
[to be squashed] Only accept multi-update digest calls when supported by cryptodev

levitte commented:
Ok, reimplemented to only allow a single update when `COP_FLAG_UPDATE` or `COP_FLAG_FINAL` aren't available.

mattcaswell commented:
> Ok, reimplemented to only allow a single update when COP_FLAG_UPDATE or COP_FLAG_FINAL aren't available.

I think this means digests have to be off by default in this scenario. For example I think libssl will break otherwise.

levitte commented:
> I think this means digests have to be off by default in this scenario. For example I think libssl will break otherwise.

Good point, and confirmed by several tests going belly up when I added this in the code:

``` C
#undef COP_FLAG_UPDATE
#undef COP_FLAG_FINAL
```

(I'll verify again that it works properly when they are defined)

levitte committed:
[to be squashed] use EVP_CIPHER_CTX_nid rather than EVP_CIPHER_CTX_type

The latter requires objects with OIDs, and some of the algorithms are not that king

levitte committed:
[to be squashed] For ciphers, flag that the default ASN1 param routines should be used

Some modes requires something different, but they are currently unimplemented

levitte committed:
[to be squashed] when preparing ciphers and digests, check that they are actually available

levitte committed:
[to be squashed] do not support digests unless COP_FLAG_UPDATE and COP_FLAG_FINAL are defined

levitte commented:
Ok, digests only available when `COP_FLAG_UPDATE` and `COP_FLAG_FINAL` are defined.  Additionally, I added run-time availability checks.  Also, a couple of bug fixes.

Tests fine on Linux now.

mattcaswell commented on crypto/engine/eng_devcrypto.c:
```diff
@@ -0,0 +1,681 @@
+
+    if (EVP_CIPHER_CTX_iv_length(ctx) > 0) {
+        if (!EVP_CIPHER_CTX_encrypting(ctx)) {
+            unsigned char *ivptr = in + inl - EVP_CIPHER_CTX_iv_length(ctx);
```
Perhaps there should be an assert to verify that `inl >= EVP_CIPHER_CTX_iv_length()`? I don't think we should have been called if this isn't the case...but that logic is not obvious from looking here.

levitte replied:
[this code](https://github.com/openssl/openssl/blob/master/crypto/evp/evp_enc.c#L320-L323) guarantees that `do_cipher` will be called with `inl` being greater than zero, and [this code](https://github.com/openssl/openssl/blob/master/crypto/evp/evp_enc.c#L320-L323) guarantees that `inl` will be an exact multiple of the block size.

So it really comes down to trusting what has been entered in `cipher_data`

mattcaswell replied:
Your code references were both the same (copy & paste error?). Anyway - I already knew this to be the case. My point was that it's non-obvious by looking at this code...and the fact that you need to go looking in a different source file somewhere else to verify that just demonstrates the point. The immediate question when looking at this code is - what happens if `EVP_CIPHER_CTX_iv_length() > inl`? Even using your logic doesn't entirely answer the question - what happens if the IV len > block size? There are a number of ciphers where `ctx->cipher->block_size == 1` for example (consider CTR mode). As it turns out that doesn't apply to any of the ciphers supported by this engine (but I had to go check your `cipher_data` definition to confirm that). An assertion would document our underlying expectation that for this situation we always expect `inl >= EVP_CIPHER_CTX_iv_length()` and prevent people spending needless cycles verifying that fact. It's also a sufficiently non-trivial assertion that you could imagine some future bug might cause it to fail (e.g. adding future support for CTR mode).

levitte replied:
Should we apply the same kind of checks in `crypto/evp/e_*.c`?  The situation is exactly the same there...  but ok, a simple assert doesn't hurt.

(the second "this code" should have been [this](https://github.com/openssl/openssl/blob/master/crypto/evp/evp_enc.c#L329))

mattcaswell replied:
> Should we apply the same kind of checks in crypto/evp/e_*.c?

Anywhere where we make non-trivial assumptions about our state is probably a candidate for considering a (soft) assert. So to answer your question - yes! But not this PR!!

mattcaswell commented on crypto/engine/eng_devcrypto.c:
```diff
@@ -0,0 +1,681 @@
+        unsigned char *ivptr = saved_iv;
+
+        if (!EVP_CIPHER_CTX_encrypting(ctx))
+            ivptr = out + inl - EVP_CIPHER_CTX_iv_length(ctx);
```
Similarly here

levitte replied:
Same answer

mattcaswell commented on crypto/engine/eng_devcrypto.c:
```diff
@@ -0,0 +1,681 @@
+                                            digest_data[i].digestlen)
+#if 0
+            || !EVP_MD_meth_set_flags(known_digest_methods[i],
+                                      digest_data[i].flags)
```
Why is this `if 0`'d out?

levitte replied:
Because I imagined that it might be needed some time in the future...  I can take it away if you insist

mattcaswell replied:
I'm not a big fan of unused code that's there "just in case". I think it should be removed.

levitte replied:
Okie

mattcaswell reviewed (changes requested):
Spotted a few more things.

levitte committed:
[to be squashed] assert that inl >= ivlen in cipher_do_cipher

levitte committed:
[to be squashed] remove unnecessary dead code

levitte commented:
All outstanding change requests answered

levitte commented:
Merged.  9a32dcf42e2f728018a02b27df9734be480b7f3c to e041f3b8baf3b9a2d57f5eb818a6b0aebd96e953

levitte commented:
Thanks!
